alloc/src/allocator.rs

```diff
@@ -0,0 +1,20 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+
+use core::alloc::{GlobalAlloc, Layout};
+use core::ffi::c_void;
+use core::ptr::null_mut;
+use mc_sgx_tservice_sys::{sgx_aligned_free, sgx_aligned_malloc};
+
+struct Allocator;
+
+unsafe impl GlobalAlloc for Allocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let size = layout.size();
+        let align = layout.align();
+        let memory = sgx_aligned_malloc(size, align, null_mut(), 0);
```

Comment:
Is this compatible with `repr(packed)`? If not, might be worth documenting. Struct alignment is something that my understanding is a bit hazy on, though, so this might be fine.

Reply:
I had to do some digging, as it wasn't specifically mentioned.
My understanding is that `repr(packed)` tells the compiler the type should be packed, no padding and alignment of 1.

Taking an example from https://notes.iveselov.info/programming/allocation-api-and-allocators#allocation-before-main
```rust
use std::alloc::{GlobalAlloc, Layout, System};

/// Allocator 
#[derive(Debug)]
pub struct What;

unsafe impl GlobalAlloc for What {
    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
        let ret = System.alloc(layout);
        return ret;
    }

    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
        println!("The layout: {layout:?}");
        System.dealloc(ptr, layout);
    }
}

#[global_allocator]
static A: What = What;
#[repr(C, packed)]
#[derive(Copy, Debug, Clone, Hash, PartialEq, Eq)]
/// some docs
pub struct Foo {
    why: u8,
    what: u32,
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn allocating_packed() {
        {
            let packed_vector: Vec<Foo>  = Vec::with_capacity(3);
            println!("the vector capacity {}", packed_vector.capacity());
        }
        assert_eq!(4, 4);

    }
}

```

Running this results with the following output
```shell
The layout: Layout { size: 112, align: 1 (1 << 0) }
, "stdout": "the vector capacity 3\nThe layout: Layout { size: 15, align: 1 (1 << 0) }\nThe layout: Layout { size: 8, align: 8 (1 << 3) }\n"The layout: Layout { size: 224, align: 1 (1 << 0) }
 }
The layout: Layout { size: 123, align: 1 (1 << 0) }
```
There is no association to types so one has to go through the debugger to see that the 15, 1 is for the vector.
![image](https://user-images.githubusercontent.com/101757619/204328088-1ea63b87-e19a-4907-95ca-0f76bb25c6da.png)